### PR TITLE
Build Caffe2 shared library on Windows

### PR DESCRIPTION
**Pull request by mingzhe09088**
Create this PR is to resolve Caffe2 shared library build issues on Windows machines. 

@orionr @Yangqing

### CODE REVIEW

**Comment by nicksterlingtek** [comment]
Thank you for opening this thread. I have been trying for several days to build on Windows 10 Pro with VS2017. Would anyone please be able to point me in the right direction? I would appreciate it so immensely. My commands were as follows.....

set "VS150COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build"
set CMAKE_GENERATOR=Visual Studio 15 2017 Win64
set DISTUTILS_USE_SDK=1
REM The following two lines are needed for Python 2.7, but the support for it is very experimental.
set MSSdk=1
set FORCE_PY27_BUILD=1
REM As for CUDA 8, VS2015 Update 3 is also required to build PyTorch. Use the following two lines.
set "PREBUILD_COMMAND=%VS140COMNTOOLS%\..\..\VC\vcvarsall.bat"
set PREBUILD_COMMAND_ARGS=x64

call "%VS150COMNTOOLS%\vcvarsall.bat" x64 -vcvars_ver=14.15
python setup.py install

OUTPUT
---------------------------------------------------------------

...

C:/Development/workspaces/repos/pytorch/aten/src/ATen/native/cuda/SpectralOps.cu(288): warning : dllexport/dllimpo
rt conflict with "at::native::_fft_cufft" [C:\Development\workspaces\repos\pytorch\build\caffe2\caffe2_gpu.vcxproj]


**"c:\Development\workspaces\repos\pytorch\build\INSTALL.vcxproj" (default target) (1) ->
"C:\Development\workspaces\repos\pytorch\build\ALL_BUILD.vcxproj" (default target) (3) ->
"C:\Development\workspaces\repos\pytorch\build\caffe2\apply_test.vcxproj" (default target) (4) ->
"C:\Development\workspaces\repos\pytorch\build\caffe2\caffe2_gpu.vcxproj" (default target) (18) ->
(CustomBuild target) ->
  C:/Development/workspaces/repos/pytorch/aten/src\ATen/core/ArrayRef.h(152): error : cannot use an entity undefined
 in device code [C:\Development\workspaces\repos\pytorch\build\caffe2\caffe2_gpu.vcxproj]
  C:/Development/workspaces/repos/pytorch/aten/src\ATen/core/ArrayRef.h(152): error : identifier "Typeinfo for  ::at
::Error" is undefined in device code [C:\Development\workspaces\repos\pytorch\build\caffe2\caffe2_gpu.vcxproj]
  Error : Internal Compiler error (codegen): "there was an error in verifying the lgenfe output!" [C:\Development\wo
rkspaces\repos\pytorch\build\caffe2\caffe2_gpu.vcxproj]**

    11950 Warning(s)
    3 Error(s)

Time Elapsed 01:56:37.06

(py36) c:\Development\workspaces\repos\pytorch\build>IF ERRORLEVEL 1 exit 1
Failed to run 'tools\build_pytorch_libs.bat --use-cuda --use-nnpack caffe2 libshm_windows'

**Comment by nicksterlingtek** [comment]
Forgot to mention I used python 3.6 in anaconda environment. Thank you!

**Comment by orionr** [comment]
@nicksterlingtek, look forward to getting this working. It might take a little bit, but definitely a top priority. :) Also feel free to follow along on https://github.com/pytorch/pytorch/pull/10765

File: caffe2/core/tensor.h [hunk 1]
```diff
@@ -952,16 +952,11 @@ class CAFFE2_API UndefinedTensorImpl final : public TensorImpl {
  //  error: identifier "at::UndefinedTensor::_singleton" is undefined in device code
  // (ostensibly because the constexpr tricks MSVC into trying to compile this
  // function for device as well).
-#ifdef _WIN32
```
Comment (peterjc123) [inline caffe2/core/tensor.h]: After looking at the comments [here](https://stackoverflow.com/questions/2479784/exporting-static-data-in-a-dll#comment2471695_2479944), I think we may have to move the implementation of the function `singleton` into the source file `tensor.cc`.

**Comment by peterjc123** [comment]
Any updates on this PR? The current build failure was addressed in the recent Windows PRs.

**Comment by orionr** [comment]
This code landed as a part of https://github.com/pytorch/pytorch/pull/11266 so closing.

**Comment by orionr** [comment]
@peterjc123 can you check if your build failures are still an issue with master now?

**Comment by peterjc123** [comment]
@orionr The build failure https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-build/13814/console seems not resolved. PR #11367 is on this.

**Pull request closed without merging**